neo/Consensus/ConsensusService.cs

```diff
@@ -225,9 +274,306 @@ private void OnPersistCompleted(Block block)
         {
             Log($"persist block: {block.Hash}");
             block_received_time = TimeProvider.Current.UtcNow;
+            knownHashes.Clear();
             InitializeConsensus(0);
         }
 
+        private bool ReverifyPrepareRequest(ConsensusContext consensusContext, RecoveryMessage message, Snapshot snapshot,
+            out ConsensusPayload prepareRequestPayload, out PrepareRequest prepareRequest)
+        {
+            if (message.TransactionHashes != null)
+            {
+                prepareRequest = new PrepareRequest
+                {
+                    ViewNumber = consensusContext.ViewNumber,
+                    Nonce = message.Nonce,
+                    NextConsensus = message.NextConsensus,
+                    TransactionHashes = message.TransactionHashes,
+                    MinerTransaction = message.MinerTransaction
+                };
+                var prepareRequestTimestamp = message.PrepareTimestamps[consensusContext.PrimaryIndex];
+                prepareRequestPayload = consensusContext.RegenerateSignedPayload(
+                    prepareRequest, (ushort) consensusContext.PrimaryIndex,
+                    message.PrepareWitnessInvocationScripts[consensusContext.PrimaryIndex], prepareRequestTimestamp);
+
+                if (prepareRequestPayload.Verify(snapshot) && PerformBasicConsensusPayloadPreChecks(prepareRequestPayload))
+                    return true;
+            }
+
+            prepareRequestPayload = null;
+            prepareRequest = null;
+            return false;
+        }
+
+        private void HandleRecoveryInCurrentView(RecoveryMessage message, Snapshot snap)
+        {
+            // If we are already on the right view number we want to accept more preparation messages and also accept
+            // any commit signatures in the payload.
+
+            if (!context.State.HasFlag(ConsensusState.CommitSent))
+            {
+                UInt256 preparationHash;
+                if (context.PrimaryIndex == context.MyIndex)
+                {
+                    preparationHash = context.Preparations[context.PrimaryIndex];
+
+                    if (preparationHash == null)
+                    {
+                        // TODO
+                        // If we haven't sent the prepare request and we are the primary. Then we can should accept
+                        // our own previously generated prepare request here if it is present in the recovery message.
```

Comment:
This is the last thing I'm expecting to address before I start testing again.

Reply:
Accepting previously sent prepare request is tricky because if for some reason it can't accept it's previous prepare request; should it generate a new one; or should it change view?
If it accepts its previous one it needs to behave completely like a backup, but if something goes wrong then it will end up switching its state to `ViewChanging`.

Reply:
A little tricky, you are right.

Did you see any case in which this PrepareRequest was not accept due to normal errors? From our perspective it is only with malicious attempt.
Thus, the whole recover should be discarded and that incident reported.

Reply:
Another node cannot maliciously sign and pretend to give a prepare request for the view that is the primary.

Reply:
maybe I need to verify the commit payload also for the primary

Reply:
Sorry i edited and re-edited

Reply:
I decided to make it completely chose to be a backup. In this case if something goes wrong it will end up changing view. The request that it accepts is it's own request though in this case only; so it will never be a means of attacking the network. Unless it wanted to attack itself... which isn't really an attack; it would just be like it decided to not create a block during it's turn and require everyone wait to change view.